### ISSUE TITLE
remove should be optimized

### ISSUE THREAD

**Issue by planetis-m** [planetis-m/packedjson2#25]

**Comment by planetis-m** [comment]
With JsonPtr we walk the tree from the root, since we have information about every parent node, there is an opportunity for a more optimized implementation based on prepare/patch

**Comment by planetis-m** [comment]
interesting benchmark

```
posFromPtr:
  Collected 1000000 samples in 2.3879 s
  Average time: 0.0016 ms
  Stddev  time: 0.0002 ms
  Min     time: 0.0001 ms
  Max     time: 0.0207 ms
270
traverse:
  Collected 1000000 samples in 1.9752 s
  Average time: 0.0012 ms
  Stddev  time: 0.0001 ms
  Min     time: 0.0000 ms
  Max     time: 0.0255 ms
270
parent:
  Collected 1000000 samples in 1.9107 s
  Average time: 0.0011 ms
  Stddev  time: 0.0000 ms
  Min     time: 0.0000 ms
  Max     time: 0.0051 ms
```

Working backwards is as slow as the opposite with all those string comparisons but faster than splitting the string.

**Comment by planetis-m** [comment]
fixed in 472561e5dd28f88fc9945f488872055fb26711c5 . It need a benchmarks.

**Comment by planetis-m** [comment]
Code that got discarded, it used an iterator

```nim
iterator posFromPtr(tree: JsonTree; path: string; n: NodePos; noDash = true): NodePos =
  # Warning: must break from the iterator manually when nilNodeId is returned.
  var n = n
  yield n
  var cur = ""
  var last = 1
  while last <= len(path):
    var first = last
    while last < len(path) and path[last] != '/':
      inc(last)
    let last1 = last == len(path)
    copyTokenToBuffer(cur, path, first, last)
    case n.kind
    of opcodeObject:
      unescapeJsonPtr(cur)
      n = rawGet(tree, n, cur)
      if not n.isNil: yield NodePos(n.int-2)
      yield n
    of opcodeArray:
      var i = getArrayIndex(cur)
      if i == -1:
        if not noDash and last1: yield NodePos(n.int+n.operand); break
        else: raiseSyntaxError(path)
      block searchLoop:
        for x in sonsReadonly(tree, n):
          if i == 0:
            n = x
            break searchLoop
          dec i
        yield nilNodeId
    else: yield nilNodeId
    inc(last)

proc remove*(tree: var JsonTree, path: JsonPtr) =
  ## Removes `path`.
  var insertPos: seq[PatchPos] = @[]
  for x in posFromPtr(tree, path.string, rootNodeId):
    if x.isNil: break
    insertPos.add x.PatchPos
  var n = insertPos.pop().int
  if insertPos.len > 0 and kind(NodePos insertPos[^1]) == opcodeKeyValuePair:
    n = insertPos.pop().int
  let diff = span(tree, n).int32
  while insertPos.len > 0:
    let pos = insertPos.pop().int
    let distance = tree.nodes[pos].operand - diff
    tree.nodes[pos] = toNode(tree.nodes[pos].kind, distance)
  let oldfull = tree.nodes.len
  for i in countup(n, oldfull-diff-1): tree.nodes[i] = tree.nodes[i+diff]
  setLen(tree.nodes, oldfull-diff)

proc posFromPtr(tree: JsonTree; path: string): NodePos =
  result = nilNodeId
  for x in posFromPtr(tree, path.string, rootNodeId):
    result = x
    if x.isNil: break
```
